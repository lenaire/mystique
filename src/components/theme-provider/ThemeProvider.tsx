import { FC } from "react";
import {
  DefaultTheme, // This is overridden in types/theme.d.ts
  ThemeProvider as StyledComponentsThemeProvider,
  ThemeProviderProps as StyledComponentsThemeProviderProps,
} from "styled-components";

import { defaultTheme } from "$cmp/theme-provider/theme";
import { ResetGlobalCSS } from "$cmp/theme-provider/ResetGlobalCSS";

export type ThemeProviderProps = Partial<StyledComponentsThemeProviderProps<DefaultTheme>>;

export const ThemeProvider: FC<ThemeProviderProps> = ({ children, theme = defaultTheme }) => {
  return (
    <StyledComponentsThemeProvider theme={theme}>
      <ResetGlobalCSS />
      {children}
    </StyledComponentsThemeProvider>
  );
};
