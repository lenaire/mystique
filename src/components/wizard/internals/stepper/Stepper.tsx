import { FC, useContext } from "react";
import { BareStepper, BareStepType } from "$cmp/wizard/internals/bare-stepper/BareStepper";
import WizardContext from "$cmp/wizard/internals/context/WizardContext";
import styled from "styled-components";
import { CmpPropsWithChildren } from "$types";

const Container = styled.div``;

export interface StepperProps extends CmpPropsWithChildren {
  steps: BareStepType;
  nonLinear?: boolean;
}

export const Stepper: FC<StepperProps> = ({ nonLinear = false, className, steps }) => {
  const { activeStep, completed } = useContext(WizardContext);
  return (
    <Container className={className}>
      <BareStepper
        activeStep={activeStep as number}
        nonLinear={nonLinear}
        steps={steps}
        completed={completed as boolean}
      />
    </Container>
  );
};
