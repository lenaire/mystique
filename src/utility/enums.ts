export enum Direction {
    Up = "up",
    Down = "down",
    Left = "left",
    Right = "right",
  }
  
  export enum Position {
    TopRight = "top-right",
    TopLeft = "top-left",
    BottomRight = "bottom-right",
    BottomLeft = "bottom-left",
  }
  
  export enum NotificationType {
    Info = "info",
    Success = "success",
    Warning = "warning",
    Error = "error",
  }