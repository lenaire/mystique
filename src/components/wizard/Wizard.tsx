import React, { FC, useMemo } from "react";
import styled from "styled-components";
import { CmpPropsWithChildren } from "$types";
import { Stepper } from "$cmp/wizard/internals/stepper/Stepper";
import { Step } from "$cmp/wizard/internals/step/Step";
import { WizardHeader } from "$cmp/wizard/internals/wizard-header/WizardHeader";
import { WizardFooter } from "$cmp/wizard/internals/wizard-footer/WizardFooter";
import WizardContext from "$cmp/wizard/internals/context/WizardContext";

const Container = styled.div``;

const Section = styled.div``;

export interface WizardProps extends CmpPropsWithChildren, React.HTMLAttributes<HTMLDivElement> {
  activeStep: number;
  completed?: boolean;
  totalSteps: number;
}

export interface WizardComponent extends FC<WizardProps> {
  Stepper: typeof Stepper;
  Step: typeof Step;
  Header: typeof WizardHeader;
  Footer: typeof WizardFooter;
}

export const Wizard: WizardComponent = ({
  activeStep,
  className,
  completed,
  totalSteps,
  children,
  ...rest
}) => {
  const contextValue = useMemo(
    () => ({
      activeStep,
      completed,
      totalSteps,
    }),
    [activeStep, completed, totalSteps]
  );

  return (
    <WizardContext.Provider value={contextValue}>
      <Container className={className} {...rest}>
        <Section>{children}</Section>
      </Container>
    </WizardContext.Provider>
  );
};

Wizard.Stepper = Stepper;
Wizard.Step = Step;
Wizard.Header = WizardHeader;
Wizard.Footer = WizardFooter;
