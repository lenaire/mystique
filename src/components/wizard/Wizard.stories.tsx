import { ComponentStory, ComponentMeta } from "@storybook/react";
import { Wizard } from "$cmp/wizard/Wizard";
import { useArgs } from "@storybook/client-api";
import styled from "styled-components";
import { Direction } from "$utility/enums";
import { Button } from "$cmp/button/Button";

// Test Build
// TODO:  Accessibility
// Storybook builds integration
// Deploy first version

export default {
  title: "Wizard",
  component: Wizard,
  argTypes: {
    className: { control: { disable: true } },
  },
} as ComponentMeta<typeof Wizard>;

const descriptionOne = `Lorem Ipsum is simply dummy text of the printing and typesetting industry. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged.`;
const descriptionTwo = `Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.`;

const Stepper = styled(Wizard.Stepper)`
  width: 300px;
`;

const Footer = styled(Wizard.Footer)`
  display: flex;
  justify-content: space-between;
  padding-top: 20px;
`;

const Body = styled.div`
  width: 600px;
`;

const StepContent = styled.div`
  height: 90%;
`;

const Template: ComponentStory<typeof Wizard> = (args) => {
  const [{ activeStep }, updateArgs] = useArgs();
  return (
    <Wizard {...args}>
      <div style={{ display: "flex" }}>
        <Wizard.Header>
          <Stepper
            steps={[
              { label: "Step 1", description: descriptionOne },
              { label: "Step 2", description: descriptionTwo },
              { label: "Step 3", description: "Completed" },
            ]}
          />
        </Wizard.Header>
        <Body>
          <StepContent>
            <Wizard.Step index={1} transition direction={Direction.Left}>
              <p>Step 1</p>
            </Wizard.Step>
            <Wizard.Step index={2} transition direction={Direction.Left}>
              <p>Step 2</p>
            </Wizard.Step>
            <Wizard.Step index={3} transition direction={Direction.Left}>
              <p>Step 3</p>
            </Wizard.Step>
          </StepContent>
          <Footer>
            <Button
              type="primary"
              label="Prev"
              onClick={(): void => updateArgs({ activeStep: activeStep - 1 })}
            />
            <Button
              type="primary"
              label="Next"
              onClick={(): void => {
                updateArgs({ activeStep: activeStep + 1 });
              }}
            />
          </Footer>
        </Body>
      </div>
    </Wizard>
  );
};

export const Primary = Template.bind({});
Primary.args = {
  activeStep: 1,
  totalSteps: 3,
  completed: false,
};
