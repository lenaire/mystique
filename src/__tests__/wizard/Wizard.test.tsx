import renderer from "react-test-renderer";
import { ThemeProvider } from "$cmp/theme-provider/ThemeProvider";
import { Wizard, WizardProps } from "$cmp/wizard/Wizard";
import { Stepper } from "$cmp/wizard/internals/stepper/Stepper";
import { Direction } from "$utility/enums";

describe("<Toaster />", () => {
  it("renders correctly", () => {
    const props: WizardProps = {
      activeStep: 1,
      totalSteps: 2,
      completed: false,
    };

    const tree = renderer
      .create(
        <ThemeProvider>
          <Wizard {...props}>
            <div style={{ display: "flex" }}>
              <Wizard.Header>
                <Stepper
                  steps={[
                    { label: "Step 1", description: "Step 1" },
                    { label: "Step 2", description: "Step 2" },
                    { label: "Step 3", description: "Completed" },
                  ]}
                />
              </Wizard.Header>
              <Wizard.Step index={1} transition direction={Direction.Left}>
                <p>Step 1</p>
              </Wizard.Step>
              <Wizard.Step index={2} transition direction={Direction.Left}>
                <p>Step 2</p>
              </Wizard.Step>
              <Wizard.Step index={3} transition direction={Direction.Left}>
                <p>Step 3</p>
              </Wizard.Step>
            </div>
            <Wizard.Footer>
              Footer
            </Wizard.Footer>
          </Wizard>
        </ThemeProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
