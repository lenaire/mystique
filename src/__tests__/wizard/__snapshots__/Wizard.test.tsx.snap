// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toaster /> renders correctly 1`] = `
<div
  className="sc-csuSiG"
>
  <div
    className="sc-eDWCr"
  >
    <div
      style={
        {
          "display": "flex",
        }
      }
    >
      <div
        className="sc-kDvujY"
      >
        <div
          className="sc-pyfCe"
        >
          <div
            className="sc-hLBbgP"
          >
            <div
              className="sc-eDvSVe"
            >
              <div
                className="sc-gKPRtg kJaZZ"
                onClick={[Function]}
              >
                <div
                  className="sc-bcXHqe KpOAM"
                  title="ActiveStep"
                />
                <span
                  className="sc-iBYQkv gVhsnL"
                >
                  Step 1
                </span>
              </div>
              <div
                className="sc-ftTHYK bmqxSp"
              >
                <p
                  className="sc-jSUZER eccVIr"
                >
                  Step 1
                </p>
              </div>
            </div>
            <div
              className="sc-eDvSVe"
            >
              <div
                className="sc-gKPRtg kJaZZ"
                onClick={[Function]}
              >
                <div
                  className="sc-dkrFOg dgYQmO"
                  title="IncompleteStep"
                />
                <span
                  className="sc-iBYQkv gWXaFx"
                >
                  Step 2
                </span>
              </div>
              <div
                className="sc-ftTHYK eHXvhP"
              >
                <p
                  className="sc-jSUZER eccVIr"
                >
                  Step 2
                </p>
              </div>
            </div>
            <div
              className="sc-eDvSVe"
            >
              <div
                className="sc-gKPRtg kJaZZ"
                onClick={[Function]}
              >
                <div
                  className="sc-dkrFOg dgYQmO"
                  title="IncompleteStep"
                />
                <span
                  className="sc-iBYQkv gWXaFx"
                >
                  Step 3
                </span>
              </div>
              <div
                className="sc-ftTHYK elUIpr"
              >
                <p
                  className="sc-jSUZER eccVIr"
                >
                  Completed
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="sc-jrcTuL eNOLMQ"
        direction="left"
      >
        <p>
          Step 1
        </p>
      </div>
      <div
        className="sc-jrcTuL knzFJL"
        direction="left"
      >
        <p>
          Step 2
        </p>
      </div>
      <div
        className="sc-jrcTuL knzFJL"
        direction="left"
      >
        <p>
          Step 3
        </p>
      </div>
    </div>
    <div
      className="sc-ipEyDJ"
    >
      Footer
    </div>
  </div>
</div>
`;
